### ISSUE TITLE
Ray + Horovod : hvd.init() fails on MacOS

### ISSUE THREAD

**Issue by Deathn0t** [horovod/horovod#2693]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.2.0
3. Horovod version: 0.21.3
4. MPI version: mpirun (Open MPI) 4.0.5
5. CUDA version: None
6. NCCL version: None
7. Python version: 3.7.9
8. Spark / PySpark version: None
9. OS and version: macOS Big Sur 11.2
10. GCC version: Apple clang version 12.0.0 (clang-1200.0.32.29)
11. CMake version: 3.19.3
12. Ray version: 1.2.0

**Bug report:**

While trying to run the [examples/ray/tensorflow2_mnist_ray.py](https://github.com/horovod/horovod/blob/master/examples/ray/tensorflow2_mnist_ray.py) with `python tensorflow2_mnist_ray.py` the following errors appears:

```
2021-02-25 08:46:26,142	INFO services.py:1174 -- View the Ray dashboard at http://127.0.0.1:8265
(pid=8582) [/private/var/folders/20/45pr4b1n2vj8rpt6vys5wk5h0000gn/T/pip-install-bbp3loh7/horovod_e32d73fa0308444b8a908c32a4c684f7/third_party/gloo/gloo/transport/uv/libuv.h:602] uv_read_start: connection already in progress
2021-02-25 08:46:29,995	WARNING worker.py:1107 -- A worker died or was killed while executing task ffffffffffffffff19549ad93c7787cd0ef95f6801000000.
Traceback (most recent call last):
  File "tensorflow2_example.py", line 104, in <module>
    executor.run(train, kwargs=dict(num_epochs=1))
  File "/Users/romainegele/opt/anaconda3/envs/dhtest/lib/python3.7/site-packages/horovod/ray/runner.py", line 433, in run
    return ray.get(self.run_remote(fn, args, kwargs))
  File "/Users/romainegele/opt/anaconda3/envs/dhtest/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
  File "/Users/romainegele/opt/anaconda3/envs/dhtest/lib/python3.7/site-packages/ray/worker.py", line 1458, in get
    raise value
ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task. Check python-core-worker-*.log files for more information.
```

The following example also fails:
```
import ray

# import horovod.tensorflow.keras as hvd
import tensorflow as tf
import horovod.tensorflow as hvd
from horovod.ray import RayExecutor

# Start the Ray cluster or attach to an existing Ray cluster
ray.init(num_cpus=2)

# Ray executor settings
setting = RayExecutor.create_settings(timeout_s=100)
num_hosts = 1  # number of machine to use
num_slots = 2  # number of workers to use on each machine
cpus_per_slot = 1  # number of cores to allocate to each worker
gpus_per_slot = None  # number of GPUs to allocate to each worker

# Start num_hosts * num_slots actors on the cluster
# https://horovod.readthedocs.io/en/stable/api.html#horovod-ray-api
executor = RayExecutor(
    setting,
    num_hosts=num_hosts,
    num_slots=num_slots,
    # cpus_per_slot=cpus_per_slot,
    # gpus_per_slot=gpus_per_slot,
)

# Launch the Ray actors on each machine
# This will launch `num_slots` actors on each machine
executor.start()


#! Note that there is an implicit assumption on the cluster being
#! homogenous in shape (i.e., all machines have the same number of
#! slots available). This is simply an implementation detail and is
#! not a fundamental limitation.

# Using the stateless `run` method, a function can take in any args or kwargs
def simple_fn():


    hvd.init()

    print("here")
    # print("hvd rank", hvd.rank())
    # return hvd.rank()
    return 0


# Execute the function on all workers at once
result = executor.run(simple_fn)
print(result)

# Check that the rank of all workers is unique
assert len(set(result)) == num_hosts * num_slots

executor.shutdown()
```

However if the `hvd.init()` is commented out then it works fine.

**Comment by tgaddair** [comment]
Hey @richardliaw, any thoughts on this issue?

**Comment by richardliaw** [comment]
Hey Romain! Could you provide the output from `/tmp/ray/session_latest`?

**Comment by Deathn0t** [comment]
Here it is! Thank you for the help! The same test on Debian/Ubuntu succeeded!

[latest.zip](https://github.com/horovod/horovod/files/6045304/latest.zip)

**Comment by richardliaw** [comment]
Looks like 
```
[2021-02-25 10:49:17,701 I 90978 336636] direct_actor_transport.cc:441: Actor creation task finished, task_id: ffffffffffffffff3cdba30494d5b0f9d6744a5601000000, actor_id: 3cdba30494d5b0f9d6744a5601000000
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415: *** Aborted at 1614246559 (unix time) try "date -d @1614246559" if you are using GNU date ***
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415: PC: @                0x0 (unknown)
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415: *** SIGABRT (@0x7fff20461462) received by PID 90978 (TID 0x700007b3e000) stack trace: ***
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @     0x7fff204d3d7d _sigtramp
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @ 0x3934316130303934 (unknown)
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @     0x7fff203e2720 abort
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccac8e21 gloo::transport::uv::libuv::TCP::read()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaa9970 gloo::transport::uv::libuv::TCP::read()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaa8390 gloo::transport::uv::Pair::connectCallback()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaac751 std::__1::__function::__func<>::operator()()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1cca971b4 gloo::transport::uv::Device::connectAsListenerCallback()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaa32ee std::__1::__function::__func<>::operator()()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaa6f16 gloo::transport::uv::libuv::Emitter<>::Handler<>::publish()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccaa69f1 gloo::transport::uv::libuv::TCP::uv__read_cb()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccabc81e uv__stream_io
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccac35b7 uv__io_poll
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1ccab4b1e uv_run
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @        0x1cca9a7ea std::__1::__thread_proxy<>()
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @     0x7fff2048f950 _pthread_start
[2021-02-25 10:49:19,785 E 90978 337853] logging.cc:415:     @     0x7fff2048b47b thread_start
```
failed?

@tgaddair have you seen this before?

**Comment by tgaddair** [comment]
Hmm, looks like an issue with libuv.

@Deathn0t, have you run `brew install libuv`?  May also be worth upgrading if it's already installed.

**Comment by Deathn0t** [comment]
Thanks for looking at this issue.
@tgaddair yes I did, here is a few more details:
- I have `libuv--1.41.0.big_sur.bottle.tar.gz`
- Installed with `brew install libuv`
- I did `brew upgrade libuv` but looks like it is already the latest version

**Comment by tgaddair** [comment]
After reproducing the issue, I can verify that this was fixed by the Gloo update in #2798.
